packages/superset-ui-encodable/src/encoders/Encoder.ts

```diff
@@ -0,0 +1,98 @@
+import { flatMap } from 'lodash';
+import { ChannelDef, TypedFieldDef } from '../types/ChannelDef';
+import { MayBeArray } from '../types/Base';
+import { isFieldDef } from '../typeGuards/ChannelDef';
+import { isNotArray } from '../typeGuards/Base';
+import ChannelEncoder from './ChannelEncoder';
+import {
+  EncodingConfig,
+  DeriveEncoding,
+  DeriveChannelTypes,
+  DeriveChannelEncoders,
+} from '../types/Encoding';
+
+export default class Encoder<Config extends EncodingConfig> {
+  readonly encoding: DeriveEncoding<Config>;
+  readonly channelTypes: DeriveChannelTypes<Config>;
+  readonly channels: DeriveChannelEncoders<Config>;
+
+  readonly legends: {
+    [key: string]: (keyof Config)[];
+  };
+
+  constructor({
+    channelTypes,
+    encoding,
+  }: {
+    channelTypes: DeriveChannelTypes<Config>;
+    encoding: DeriveEncoding<Config>;
+  }) {
+    this.channelTypes = channelTypes;
+    this.encoding = encoding;
+    const channelNames = this.getChannelNames();
+
+    // Create channel encoders
+    const channels: { [k in keyof Config]?: MayBeArray<ChannelEncoder<ChannelDef>> } = {};
+
+    channelNames.forEach(name => {
+      const channelEncoding = encoding[name] as MayBeArray<ChannelDef>;
+      if (Array.isArray(channelEncoding)) {
+        const definitions = channelEncoding;
+        channels[name] = definitions.map(
+          (definition, i) =>
+            new ChannelEncoder({
+              channelType: 'Text',
+              definition,
+              name: `${name}[${i}]`,
+            }),
+        );
+      } else {
+        const definition = channelEncoding;
+        channels[name] = new ChannelEncoder({
+          channelType: channelTypes[name],
+          definition,
+          name: name as string,
+        });
+      }
+    });
+
+    this.channels = channels as DeriveChannelEncoders<Config>;
+
+    // Group the channels that use the same field together
+    // so they can share the same legend.
+    this.legends = {};
+    channelNames
+      .map(name => this.channels[name])
+      .forEach(c => {
+        if (isNotArray(c) && c.hasLegend() && isFieldDef(c.definition)) {
```

Comment:
At this point, the channels that are array are fields for `tooltip`. I tried to think if we want to support other channels that require legends such as `color` as array of multiple fields but that makes thing very complicated. Better create a derived field `field1,field2` in the dataset and use that derived field instead. (`vega` also has support for `formula` instead of `field` for this kind of purpose, which I haven't explored how to port)